File: Core/DatabaseCatalogue.cs
using KolibSoft.Catalogue.Core.Abstractions;
using KolibSoft.Catalogue.Core.Utils;
using Microsoft.EntityFrameworkCore;

namespace KolibSoft.Catalogue.Core;

public class DatabaseCatalogue<TItem, TFilters> : ICatalogueConnector<TItem, TFilters>
    where TItem : class, IItem, IValidatable, IUpdatable<TItem>
    where TFilters : IPageFilters, IItemFilters
{

    public DbContext DbContext { get; }
    public DbSet<TItem> DbSet { get; }
    public ICollection<string> Errors { get; } = new List<string>();
    public virtual bool Available => DbContext.Database.CanConnect();

    protected virtual IQueryable<TItem> QueryItems(IQueryable<TItem> items, TFilters filters) => items;
    protected virtual bool ValidateInsert(TItem item) => true;
    protected virtual bool ValidateUpdate(TItem item) => true;
    protected virtual bool ValidateDelete(TItem item) => true;

    public virtual Task<Result<Page<TItem>?>> PageAsync(TFilters? filters = default) => Task.Run<Result<Page<TItem>?>>(() =>
    {
        Errors.Clear();
        var items = DbSet.AsQueryable();
        if (filters?.Ids?.Any() == true) items = items.Where(x => filters.Ids.Contains(x.Id));
        if (filters != null) items = QueryItems(items, filters);
        var page = items.GetPage(filters?.PageIndex ?? 0, filters?.PageSize ?? CatalogueStatics.DefaultChunkSize);
        return page;
    });

    public virtual Task<Result<TItem?>> GetAsync(Guid id) => Task.Run<Result<TItem?>>(() =>
    {
        Errors.Clear();
        var item = DbSet.FirstOrDefault(x => x.Id == id);
        if (item == null)
        {
            Errors.Add(CatalogueStatics.NoItem);
            return Errors.ToArray();
        }
        return item;
    });

    public virtual Task<Result<Guid?>> InsertAsync(TItem item) => Task.Run<Result<Guid?>>(() =>
    {
        Errors.Clear();
        if (!item.Validate(Errors)) return Errors.ToArray();
        if (DbSet.Any(x => x.Id == item.Id))
        {
            Errors.Add(CatalogueStatics.RepeatedItem);
            return Errors.ToArray();
        }
        if (!ValidateInsert(item)) return Errors.ToArray();
        DbSet.Add(item);
        DbContext.SaveChanges();
        return item.Id;
    });

    public virtual Task<Result<bool?>> UpdateAsync(Guid id, TItem item) => Task.Run<Result<bool?>>(() =>
    {
        Errors.Clear();
        if (!item.Validate(Errors)) return Errors.ToArray();
        var original = DbSet.FirstOrDefault(x => x.Id == id);
        if (original == null)
        {
            Errors.Add(CatalogueStatics.NoItem);
            return Errors.ToArray();
        }
        if (item.UpdatedAt < original.UpdatedAt) return false;
        if (!ValidateUpdate(item)) return Errors.ToArray();
        original.Update(item);
        DbSet.Update(original);
        DbContext.SaveChanges();
        return true;
    });

    public virtual Task<Result<bool?>> DeleteAsync(Guid id) => Task.Run<Result<bool?>>(() =>
    {
        Errors.Clear();
        var item = DbSet.FirstOrDefault(x => x.Id == id);
        if (item == null)
        {
            Errors.Add(CatalogueStatics.NoItem);
            return Errors.ToArray();
        }
        if (!ValidateDelete(item)) return Errors.ToArray();
        DbSet.Remove(item);
        DbContext.SaveChanges();
        return true;
    });

    public DatabaseCatalogue(DbContext dbContext)
    {
        DbContext = dbContext;
        DbSet = DbContext.Set<TItem>();
    }

}

File: docs/lib/database_catalogue.js
import { CatalogueConnector } from "./abstractions/catalogue_connector.js";
import { CatalogueFilters } from "./catalogue_filters.js";
import { Page } from "./page.js";
import * as CatalogueStatics from "./catalogue_statics.js";
import { Result } from "./result.js";
import * as PageUtils from "./utils/page_utils.js";
import { DbContext } from "./dbcontext.js";
import { DbSet } from "./dbset.js";

/**
 * @template TItem
 * @template TFilters
 * @implements {CatalogueConnector}
 */
class DatabaseCatalogue {

    /** @type {DbContext} */
    dbContext;

    /** @type {DbSet<TItem>} */
    dbset;

    /** @type {string[]} */
    errors;

    /**
     * @param {DbContext} dbContext 
     * @param {string} name 
     */
    constructor(dbContext, name) {
        this.dbContext = dbContext;
        this.dbset = dbContext.set(name);
        this.errors = [];
    }

    /** @returns {boolean} */
    get available() { return true; }

    /**
     * @param {TItem[]} items 
     * @param {TFilters} filters 
     * @returns {TItem[]}
     */
    queryItems(items, filters) { return items; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    validateInsert(item) { return true; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    validateUpdate(item) { return true; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    validateDelete(item) { return true; }

    /**
     * @param {TFilters} filters 
     * @returns {Promise<Result<Page<TItem>>>}
     */
    async pageAsync(filters = null) {
        this.errors = [];
        let items = await this.dbset.filter(item => {
            if (filters?.changesAt && item.updatedAt < filters.changesAt) return false;
            if (filters?.ids?.length && !(filters.ids.includes(item.id))) return false;
            return true;
        });
        items = this.queryItems(items, filters);
        let page = PageUtils.getPage(items, filters?.pageIndex ?? 0, filters?.pageSize ?? CatalogueStatics.DefaultChunkSize);
        return new Result({
            data: new Page({
                items: page.items,
                pageIndex: page.pageIndex,
                pageCount: page.pageCount
            })
        });
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem>>}
     */
    async getAsync(id) {
        this.errors = [];
        let item = await this.dbset.get(id);
        return new Result({
            data: item
        });
    }

    /**
     * @param {TItem} item 
     * @returns {Promise<Result<string>>}
     */
    async insertAsync(item) {
        this.errors = [];
        if (item.validate && !item.validate(this.errors)) return new Result({ errors: this.errors });
        if (await this.dbset.get(item.id)) {
            this.errors.push(CatalogueStatics.RepeatedItem);
            return new Result({ errors: this.errors });
        }
        if (!this.validateInsert(item)) return new Result({ errors: this.errors });
        await this.dbset.add(item);
        return new Result({ data: item.id });
    }

    /**
     * @param {string} id 
     * @param {TItem} item 
     * @returns {Promise<Result<boolean>>}
     */
    async updateAsync(id, item) {
        this.errors = [];
        if (item.validate && !item.validate(this.errors)) return new Result({ errors: this.errors });
        let original = await this.dbset.get(id);
        if (original == null) {
            this.errors.push(CatalogueStatics.NoItem);
            return new Result({ errors: this.errors });
        }
        if (item.UpdatedAt < original.UpdatedAt) return new Result({ data: false });
        if (!this.validateUpdate(item)) return new Result({ errors: this.errors });
        await this.dbset.update(item);
        return new Result({ data: true });
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<boolean>>}
     */
    async deleteAsync(id) {
        this.errors = [];
        let item = this.dbset.get(id);
        if (item == null) {
            this.errors.push(CatalogueStatics.NoItem);
            return new Result({ errors: this.errors });
        }
        if (!this.validateDelete(item)) return new Result({ errors: this.errors });
        await this.dbset.remove(id);
        return new Result({ data: true });
    }

}

export {
    DatabaseCatalogue
}